### PR TITLE
SY-3964: Add Dot Based Module Syntax to Arc

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-####]()

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR adds dot-based module syntax to Arc (e.g. `math.pow(x, 2)`, `time.interval{}`, `string.concat(a, b)`) by extending the ANTLR grammar with a `QualifiedIdentifier` rule and propagating the change through the parser AST helpers, analyzer, compiler, LSP (completion, hover, semantic tokens), and both the Go and C++ WASM runtimes. It also fixes several pre-existing bugs discovered while building the feature: `Density()` panicking on string-typed named outputs, string channel data not being bridged to i32 WASM handles, `string.len/concat/equal` rejecting `str`-typed arguments, and `math.pow`'s generic output returning the wrong type.

<h3>Confidence Score: 5/5</h3>

- Safe to merge; all remaining findings are P2 or lower.
- The feature is implemented consistently across the full stack (parser → analyzer → compiler → runtime in both Go and C++), parity is maintained between Go and C++ runtimes, and the PR ships extensive unit and integration tests. The two findings are: a documentation typo in an LSP completion hint (i8 range "172" vs "127") and an edge-case where a generic function called with only constant arguments surfaces a runtime error rather than a compile-time diagnostic. Neither blocks correctness for typical Arc programs.
- arc/go/compiler/expression/compiler.go — type-variable unification edge case for all-constant generic calls; arc/go/lsp/completion.go — i8 doc typo.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| arc/go/parser/ast.go | Adds QualifiedName, FunctionName, and PrimaryName helpers to extract dot-qualified identifiers from parse contexts; cleanly handles both qualified (math.pow) and unqualified forms. |
| arc/go/compiler/expression/compiler.go | Adds generic type-variable unification across function arguments (resolving T from channels so literals get the correct concrete hint), and routes qualified identifiers through compilePrimary; logic is sound for the primary use-case but all-constant calls to generic functions leave T unresolved. |
| arc/go/stl/wasm/wasm.go | Adds unresolved-type guard (IsValid), correct 4-byte string offset accounting, and stringInputs bookkeeping passed into nodeImpl; consistently mirrors the C++ module.h changes. |
| arc/go/stl/wasm/node.go | Adds inline string-to-handle conversion for WASM string inputs; correctly uses uint32 truncation in setValueAt to discard the upper 32 bits when reading 8-byte aligned memory for 4-byte string handles. |
| arc/cpp/runtime/wasm/module.h | C++ counterpart to Go wasm.go: adds unresolved-type validation, 4-byte string offset accounting, and correct byte_size for memcpy; consistent with Go implementation. |
| arc/cpp/runtime/wasm/node.h | C++ counterpart to Go node.go: adds string_inputs vector and str_state, inline string-to-handle conversion using at<std::string>; mirrors Go logic faithfully. |
| arc/go/stl/strings/string.go | Updates user-facing string function signatures from raw i32 to the String type, adds KindFunction kind to all members, and fixes output param names to use ir.DefaultOutputParam; internal from_literal correctly retains Internal flag. |
| arc/go/stl/time/time.go | Adds "time.interval" / "time.wait" qualified-name cases to the factory switch with an explanatory comment; the acknowledged workaround is safe and inexpensive. |
| arc/go/lsp/completion.go | Adds module-qualified prefix completion (math., time.) with FilterText/TextEdit and KindFunction-only filtering; also contains a pre-existing typo in the i8 doc range (172 vs 127). |
| arc/go/compiler/statement/variable.go | Fixes Density() panic on string-typed named outputs by explicitly handling KindString (4-byte i32 store) in both the offset calculation and the store-op switch. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User writes math.pow&#40;ch, 2&#41;] --> B[ANTLR Parser\nQualifiedIdentifier rule]
    B --> C[AST helpers\nQualifiedName / PrimaryName]
    C --> D[Analyzer\nanalyzePrimary / inferPrimaryType]
    C --> E[Compiler\ncompilePrimary / compileFunctionCallExpr]
    E --> F[Type Variable Unification\nvarMap resolves T from channel arg]
    F --> G[IR EmitCall\nconcreteType with resolved T]
    G --> H[WASM node creation\nIsValid guard / string offset accounting]
    H --> I[Go wasm.nodeImpl.Next\nstringInputs handles]
    H --> J[C++ wasm::Node::next\nstring_inputs handles]

    C --> K[LSP Completion\nmodule-qualified TextEdit]
    C --> L[LSP Hover\nGetQualifiedWordAtPosition]
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `arc/go/lsp/completion.go`, line 420-437 ([link](https://github.com/synnaxlabs/synnax/blob/193d0c4067af3ebde0ac04359edcad4f56e8e7d3/arc/go/lsp/completion.go#L420-L437)) 

   <a href="#">P2</a> **Module-member completions return short names with a qualified prefix**

   When the user types `math.p`, the `prefix` variable is `"math.p"` (because the dot is now treated as a word character). The `Search` call on the resolver is made with that full prefix. If the resolver correctly returns module members (e.g. `pow`) with their short names, the completion item `Label` is `"pow"` with no `textEdit` range specified.

   Most LSP clients (e.g. VSCode) determine the replacement range by the word boundary at the cursor. Since `.` is **not** a standard word boundary character in editors, the client may determine the replacement range as only `"p"` (after the last `.`), and insert `pow` over that, leaving `math.pow` — which is the desired result. However, this is editor-dependent behaviour; a client that uses the whole `"math.p"` prefix as the replacement text would produce just `"pow"`.

   Consider specifying a `TextEdit` on the completion item with an explicit replacement range covering only the member portion (after the last `.`), or splitting the prefix and member before calling `Search`, so the label is unambiguous regardless of client behaviour.
</details>

<sub>Reviews (2): Last reviewed commit: ["SY-3964: fix time.interval{} and time.wa..."](https://github.com/synnaxlabs/synnax/commit/c1362e717a7f99b3132069e18cdf1c1bb6120941) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26177080)</sub>

### CODE REVIEW

File: arc/go/parser/ast.go [hunk 1]
```diff
@@ -208,6 +208,35 @@ func GetPrimaryExpression(expr IExpressionContext) IPrimaryExpressionContext {
 	return postfix.PrimaryExpression()
 }
 
+// QualifiedName returns the dot-joined name from a qualified identifier
+// (e.g., "math.pow").
+func QualifiedName(qid IQualifiedIdentifierContext) string {
+	ids := qid.AllIDENTIFIER()
+	return ids[0].GetText() + "." + ids[1].GetText()
+}
```
[the remark this answers is not in the document]
Comment (sy-nico) [reply arc/go/parser/ast.go:216]: Verified manually that there is no issue. ANTLR's error recovery always inserts a synthetic `<missing undefined>` token, so `AllIDENTIFIER()` always returns exactly 2 elements even on incomplete input like math.. The index access doesn't panic.

Additionally, `ParseBlock` returns nil when there are parse errors (parser.go:114-117), so `QualifiedName` is never called on malformed input through the analyzer/compiler/LSP analysis paths. The user sees a parse error diagnostic ("missing IDENTIFIER at '}'") which is the correct behavior.

No change needed.

**Comment by sy-nico** [comment]
@greptile, please review and update the summary score. Your only concern for `ast.go` is a non issue. See explanation:

```
Verified manually that there is no issue. ANTLR's error recovery always inserts a synthetic `<missing undefined>` token, so `AllIDENTIFIER()` always returns exactly 2 elements even on incomplete input like math.. The index access doesn't panic.

Additionally, `ParseBlock` returns nil when there are parse errors (parser.go:114-117), so `QualifiedName` is never called on malformed input through the analyzer/compiler/LSP analysis paths. The user sees a parse error diagnostic ("missing IDENTIFIER at '}'") which is the correct behavior.

No change needed.
```

In additional, various improvements and bug-fixes have been implemented. Each STL and function has extensive testing across all layers and across all possible parameters:
* Analyzer
* Compiler
* Go/C++ Runtime
* Integration/ (end-to-end testing)
